ktor-client/ktor-client-darwin/darwin/src/io/ktor/client/engine/darwin/DarwinWebsocketSession.kt

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright 2014-2022 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.client.engine.darwin
+
+import io.ktor.client.request.*
+import io.ktor.http.*
+import io.ktor.util.date.*
+import io.ktor.utils.io.core.*
+import io.ktor.websocket.*
+import kotlinx.cinterop.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.*
+import platform.Foundation.*
+import platform.darwin.*
+import kotlin.coroutines.*
+
+@OptIn(UnsafeNumber::class, DelicateCoroutinesApi::class, ExperimentalCoroutinesApi::class)
+internal class DarwinWebsocketSession(
+    private val requestTime: GMTDate,
+    coroutineContext: CoroutineContext
+) : WebSocketSession {
+
+    internal lateinit var task: NSURLSessionWebSocketTask
+
+    internal val response = CompletableDeferred<HttpResponseData>()
+
+    private val _incoming = Channel<Frame>(Channel.UNLIMITED)
+    private val _outgoing = Channel<Frame>(Channel.UNLIMITED)
+    private val socketJob = Job(coroutineContext[Job])
+    override val coroutineContext: CoroutineContext = coroutineContext + socketJob
+
+    override var masking: Boolean
+        get() = true
+        set(_) {}
+
+    override var maxFrameSize: Long
+        get() = task.maximumMessageSize.convert()
+        set(value) {
+            task.setMaximumMessageSize(value.convert())
+        }
+
+    override val incoming: ReceiveChannel<Frame>
+        get() = _incoming
+
+    override val outgoing: SendChannel<Frame>
+        get() = _outgoing
+
+    override val extensions: List<WebSocketExtension<*>>
+        get() = emptyList()
+
+    internal val delegate: NSURLSessionWebSocketDelegateProtocol = WebSocketDelegate()
+
+    init {
+        launch {
+            receiveMessages()
+        }
+        launch {
+            sendMessages()
+        }
+        this.coroutineContext[Job]!!.invokeOnCompletion { cause ->
+            if (cause != null) {
+                val code = CloseReason.Codes.INTERNAL_ERROR.code.convert<NSInteger>()
+                task.cancelWithCloseCode(code, "Client failed".toByteArray().toNSData())
+            }
+            _incoming.close()
+            _outgoing.cancel()
+        }
+    }
+
+    private suspend fun receiveMessages() {
+        while (true) {
+            val message = suspendCancellableCoroutine<NSURLSessionWebSocketMessage?> {
```

Comment:
Consider extracting this to `suspend` extension function calling `suspendCancellableCoroutine`.

so It will look like a coroutine wrapper around callback API